### PR TITLE
Refactored to move ES6 version

### PR DESCRIPTION
**Pull request by guncebektas**
As requested on https://github.com/meteor/blaze/issues/367, I started to migrate code to ES6. 
All tests are passing, please review and let me know for required changes.

### CODE REVIEW

**Comment by guncebektas** [comment]
Besides the pr, I found some exported function has no usage in the blaze like isNully, isTagEnsured, etc... should we stop exporting them.

**Comment by jankapunkt** [comment]
Thank you @guncebektas for this PR. Please note, that we first need to merge #382 before we can continue to work on the actual migrations. Once it's done I will review this one.

**Comment by jankapunkt** [comment]
@guncebektas I merged #382  to the `master` branch so you should pull it in to use it to lint your package. Please make sure there are no remaining lint errors. You can [read in the CONTRIBUTING file how to use the linter](https://github.com/meteor/blaze/blob/master/CONTRIBUTING.md#run-the-linter).

**Comment by guncebektas** [comment]
This will be hard to achieve :)

**Comment by jankapunkt** [comment]
Hey @guncebektas sorry that was a bit fast from my side. Of course there may be design decisions from back then that will conflict with the linter. For now please only to non-breaking migrations. All other issues will be discussed during review.

**Comment by jankapunkt** [comment]
Another thing is, which I mentioned in #385 is that you should please only do one package per PR. Otherwise this gets easily out of hand and becomes impossible to review. For now please keep the scope only to the two packages you worked on.

**Comment by guncebektas** [comment]
no problem. I will try to do my best.

**Comment by guncebektas** [comment]
It wasn't possible to fulfill all linting rules but I tried to minimize errors. I think a review at this point will be good. 
All tests of the project are passing. 

In addition to this, I used my version as local packages in my project and all of my e2e tests are passing.

I can trust myself and move faster after reviews. @jankapunkt

File: packages/htmljs/visitors.js [hunk 1]
```diff
@@ -11,16 +11,17 @@ import {
 } from './html';
 
 
-var IDENTITY = function (x) { return x; };
+const IDENTITY = function (x) {
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:14]: Let's do this a nice arrow-fn:

```js
const IDENTITY = x => x
```

File: packages/htmljs/visitors.js [hunk 2]
```diff
@@ -11,16 +11,17 @@ import {
...
 // _assign is like _.extend or the upcoming Object.assign.
 // Copy src's own, enumerable properties onto tgt and return
 // tgt.
-var _hasOwnProperty = Object.prototype.hasOwnProperty;
-var _assign = function (tgt, src) {
-  for (var k in src) {
-    if (_hasOwnProperty.call(src, k))
-      tgt[k] = src[k];
+const _hasOwnProperty = Object.prototype.hasOwnProperty;
+const _assign = function (tgt, src) {
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:22]: Replace all `_assign` calls with `Object.assign` and remove this function. It's deprecated.

File: packages/htmljs/visitors.js [hunk 3]
```diff
@@ -34,85 +35,89 @@ Visitor.def = function (options) {
...
   return subType;
 };
 
 Visitor.def({
-  visit: function (content/*, ...*/) {
-    if (content == null)
+  visit(content/* , ... */) {
+    if (content == null) {
       // null or undefined.
       return this.visitNull.apply(this, arguments);
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:53]: Don't use arguments. We should use spread syntax:

```js
visit(...args) {
  const [content] = args;

  if (content == null) {
    // null or undefined.
    return this.visitNull.apply(this, args);
  }

...
```

File: packages/htmljs/visitors.js [hunk 4]
```diff
@@ -34,85 +35,89 @@ Visitor.def = function (options) {
 };
 
 Visitor.extend = function (options) {
-  var curType = this;
-  var subType = function HTMLVisitorSubtype(/*arguments*/) {
+  const CurType = this;
+  const subType = function HTMLVisitorSubtype(/* arguments */) {
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:39]: Use

```js
  const subType = function HTMLVisitorSubtype(...args) {
    Visitor.apply(this, args);
  };
```

File: packages/htmljs/visitors.js [hunk 5]
```diff
@@ -34,85 +35,89 @@ Visitor.def = function (options) {
...
-  visitNull: function (nullOrUndefined/*, ...*/) {},
-  visitPrimitive: function (stringBooleanOrNumber/*, ...*/) {},
-  visitArray: function (array/*, ...*/) {},
-  visitComment: function (comment/*, ...*/) {},
-  visitCharRef: function (charRef/*, ...*/) {},
-  visitRaw: function (raw/*, ...*/) {},
-  visitTag: function (tag/*, ...*/) {},
-  visitObject: function (obj/*, ...*/) {
-    throw new Error("Unexpected object in htmljs: " + obj);
+  visitNull(nullOrUndefined/* , ... */) {
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:87]: I know they are there but never used. However, I don't know why they were added or if there should be a future use for them. We should therefor move their variables inside the comments and add an error to indicate that this is not implemented:

```js
...

visitNull(/* nullOrUndefined , ... */) {
  throw new Error('Visitor.visitNull is not implemented');
},

...

```

If future usage raises these errors we can analyze them and find a way for an implementation

File: packages/htmljs/visitors.js [hunk 6]
```diff
@@ -121,76 +126,73 @@ TransformingVisitor.def({
...
-    var newAttrs = oldAttrs;
+    const oldAttrs = attrs;
+    let newAttrs = oldAttrs;
     if (oldAttrs) {
-      var attrArgs = [null, null];
+      const attrArgs = [null, null];
       attrArgs.push.apply(attrArgs, arguments);
-      for (var k in oldAttrs) {
-        var oldValue = oldAttrs[k];
+      for (const k in oldAttrs) {
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:188]: Use `Object.getOwnPropertyNames` instead of `for..in`

```
      Object.getOwnPropertyNames(oldAttrs).forEach(k => {
        const oldValue = oldAttrs[k];
        attrArgs[0] = k;
        attrArgs[1] = oldValue;
        const newValue = this.visitAttribute.apply(this, attrArgs);
        if (newValue !== oldValue) {
          // copy on write
          if (newAttrs === oldAttrs) newAttrs = _assign({}, oldAttrs);
          newAttrs[k] = newValue;
        }
      });
```

File: packages/htmljs/visitors.js [hunk 7]
```diff
@@ -200,49 +202,57 @@ TransformingVisitor.def({
...
+export const TEXTMODE = {
+  STRING: 1,
+  RCDATA: 2,
+  ATTRIBUTE: 3,
+};
 
 export const ToTextVisitor = Visitor.extend();
 ToTextVisitor.def({
-  visitNull: function (nullOrUndefined) {
+  visitNull(nullOrUndefined) {
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:223]: Comment variable `nullOrUndefined` or remove it

File: packages/htmljs/visitors.js [hunk 8]
```diff
@@ -252,124 +262,100 @@ ToTextVisitor.def({
...
-    var children = tag.children;
+    const { tagName } = tag;
+    let { children } = tag;
 
-    var attrs = tag.attrs;
+    let { attrs } = tag;
     if (attrs) {
       attrs = flattenAttributes(attrs);
-      for (var k in attrs) {
+      for (const k in attrs) {
```
Comment (jankapunkt) [inline packages/htmljs/visitors.js:306]: Again let's use `Object.getOwnPropertyNames`

**Review by jankapunkt — CHANGES_REQUESTED** [review]

**Comment by jankapunkt** [comment]
I just focused on `htmljs` package for now. My most comments focus on using ES6 builtin functions and syntax in favour of old or depracated syntaxes. Please let's focus first only on `htmljs` before we continue to `html-tools` package.

File: packages/htmljs/html.js [hunk 9]
```diff
@@ -1,28 +1,71 @@
+export const Tag = function () {
+};
 
-export const Tag = function () {};
 Tag.prototype.tagName = ''; // this will be set per Tag subclass
 Tag.prototype.attrs = null;
 Tag.prototype.children = Object.freeze ? Object.freeze([]) : [];
 Tag.prototype.htmljsType = Tag.htmljsType = ['Tag'];
 
+export function isConstructedObject(x) {
```
[the remark this answers is not in the document]
Comment (guncebektas) [reply packages/htmljs/html.js:9]: linter gives an error if the function is called before definition. I didn't like this but the rule is a rule.

File: packages/htmljs/html.js [hunk 10]
```diff
@@ -115,124 +142,125 @@ export function isVoidElement(tagName) {
...
 
-  var isList = isArray(attrs);
-  if (isList && attrs.length === 0)
+  if (isList && attrs.length === 0) {
     return null;
+  }
+
+  const result = {};
+
+  let i = 0;
```
[the remark this answers is not in the document]
Comment (guncebektas) [reply packages/htmljs/html.js:244]: no problem for i but ```const N = (isList ? attrs.length : 1);``` in for decrease the readibilty

File: packages/htmljs/html.js [hunk 11]
```diff
@@ -115,124 +142,125 @@ export function isVoidElement(tagName) {
...
     // null or undefined
     return true;
+  }
 
   if (isArray(node)) {
     // is it an empty array or an array of all nully items?
-    for (var i = 0; i < node.length; i++)
-      if (! isNully(node[i]))
-        return false;
+    for (const item of node) if (!isNully(item)) return false;
```
[the remark this answers is not in the document]
Comment (guncebektas) [reply packages/htmljs/html.js:217]: It seems it's needed for the test however this is better
```
export function isNully(node) {
  if (node == null) {
    // null or undefined
    return true;
  }

  return isArray(node) && node.every(isNully);
}
`
``

File: packages/htmljs/html.js [hunk 12]
```diff
@@ -115,124 +142,125 @@ export function isVoidElement(tagName) {
...
-      var value = oneAttrs[name];
-      if (! isNully(value))
+  for (; i < N; i++) {
+    const oneAttrs = (isList ? attrs[i] : attrs);
+
+    if ((typeof oneAttrs !== 'object') || isConstructedObject(oneAttrs)) {
+      throw new Error(`Expected plain JS object as attrs, found: ${oneAttrs}`);
+    }
+
+    for (const name in oneAttrs) {
```
[the remark this answers is not in the document]
Comment (guncebektas) [reply packages/htmljs/html.js:254]: Tests

tinytest - spacebars-tests - template_tests - dynamic attrs
tinytest - spacebars-tests - old - template_tests - dynamic attrs

fails on this update. I will check why

File: packages/htmljs/html.js [hunk 13]
```diff
@@ -1,28 +1,71 @@
...
-      if (! isConstructedObject(attrs)) {
+      // Treat vanilla JS objects as an attributes' dictionary.
+      if (!isConstructedObject(attrs)) {
         instance.attrs = attrs;
         i++;
       } else if (attrs instanceof Attrs) {
-        var array = attrs.value;
+        const array = attrs.value;
         if (array.length === 1) {
           instance.attrs = array[0];
```
[the remark this answers is not in the document]
Comment (guncebektas) [reply packages/htmljs/html.js:70]: Is this really necessary because in elseif block we will return array so destructuring works only to remove array[0] on line line @jankapunkt

File: packages/htmljs/visitors.js (hunk 5 above)
Comment (zodern) [reply packages/htmljs/visitors.js:87]: They actually are used. The visitor could call any or all of these depending on what types are in the AST it is visiting. For example, line 53 above calls visitNull. Each type of  visitor (for example, the TransformingVisitor in this file) overrides the functions for the ast nodes that it is interested in. The functions for other node types then do nothing.

Comment (guncebektas) [reply packages/htmljs/visitors.js:79]: Really hard to understand what it is really doing

Comment (jankapunkt) [reply packages/htmljs/visitors.js:79]: > Really hard to understand what it is doing

This is mostly due to lack of documentation of the internals and applies to most of the Blaze packages.

> Each type of visitor (for example, the TransformingVisitor in this file) overrides the functions for the ast nodes that it is interested in. The functions for other node types then do nothing by default.

So this is similar to what abstract classes with uninmplemented base methods do, right?

File: packages/htmljs/html.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (jankapunkt) [reply packages/htmljs/html.js]: Yes I agree. Lets add an eslint-disable rule for this line 70 but no general disable, only this rule

File: packages/htmljs/visitors.js (hunk 5 above)
Comment (guncebektas) [reply packages/htmljs/visitors.js:79]: Yes, they are like abstract classes

File: packages/htmljs/html.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (guncebektas) [reply packages/htmljs/html.js:251]: the spacebars-tests/package template_tests does includes `es5-shim@4.8.0` not `ecmascript@0.15.1`.
should we refactor tests to continue or adding a todo and leaving it for future refactoring is good? @jankapunkt

File: packages/html-tools/utils.js [hunk 14]
```diff
@@ -33,14 +16,29 @@ const properTagCaseMap = (function (map) {
...
+  const svgCamelCaseAttributes = 'attributeName attributeType baseFrequency baseProfile calcMode clipPathUnits contentScriptType contentStyleType diffuseConstant edgeMode externalResourcesRequired filterRes filterUnits glyphRef glyphRef gradientTransform gradientTransform gradientUnits gradientUnits kernelMatrix kernelUnitLength kernelUnitLength kernelUnitLength keyPoints keySplines keyTimes lengthAdjust limitingConeAngle markerHeight markerUnits markerWidth maskContentUnits maskUnits numOctaves pathLength patternContentUnits patternTransform patternUnits pointsAtX pointsAtY pointsAtZ preserveAlpha preserveAspectRatio primitiveUnits refX refY repeatCount repeatDur requiredExtensions requiredFeatures specularConstant specularExponent specularExponent spreadMethod spreadMethod startOffset stdDeviation stitchTiles surfaceScale surfaceScale systemLanguage tableValues targetX targetY textLength textLength viewBox viewTarget xChannelSelector yChannelSelector zoomAndPan'.split(' ');
+
+  const _properAttributeCaseMap = (map => {
+    svgCamelCaseAttributes.forEach(a => {
+      map[asciiLowerCase(a)] = a;
+    });
+    return map;
+  })({});
```
Comment (zodern) [inline packages/html-tools/utils.js:40]: This function can be called very frequently. For performance, it would be better to keep lines 33 - 40 outside of the function like it was before so that work only has to be done once.

File: packages/html-tools/package.js [hunk 15]
```diff
@@ -1,8 +1,9 @@
+/* eslint-env meteor */
 Package.describe({
   name: 'html-tools',
-  summary: "Standards-compliant HTML tools",
-  version: '1.1.3',
-  git: 'https://github.com/meteor/blaze.git'
+  summary: 'Standards-compliant HTML tools',
+  version: '1.1.4',
```
Comment (StorytellerCZ) [inline packages/html-tools/package.js:5]: This should be a feature release, so `1.2.0`

**Review by StorytellerCZ — CHANGES_REQUESTED** [review]

**Comment by jankapunkt** [comment]
We are getting closer :-)

**Review by StorytellerCZ — APPROVED** [review]

**Comment by guncebektas** [comment]
Are we ok with htmljs @jankapunkt

**Comment by StorytellerCZ** [comment]
Once @jankapunkt approves, I will merge this.

**Comment by guncebektas** [comment]
Is there a prefered package for the next pr?

**Comment by jankapunkt** [comment]
hey @guncebektas @StorytellerCZ this is quite a bit to review (which is why in the future there should be only one package per PR) I try to get it done asap the next days!

**Review by jankapunkt — APPROVED** [review]
Generally approved, waiting for the comment on the package version bumps, otherwise this is a great step in the right direction

File: packages/htmljs/html.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (DanielDornhardt) [reply packages/htmljs/html.js:10]: Words are spoken... to be broken... :D -> I think it's ok to skip this linter rule in case of higher goods, eg. Reviewers comfort :D / being able to review the changes effectively. IMHO?

File: packages/htmljs/html.js
Comment (DanielDornhardt) [inline packages/htmljs/html.js:1]: Sorry, hard to parse changes

**Review by DanielDornhardt — COMMENTED** [review]
Hi.

I've reviewed this, unfortunately it's hard to review because of the amount of stylistic changes. Which is unfortunate, because this has been recommended in the past.

We added some feedback, but basically I think we're gonna start over with the refactoring and potentially pick out the good parts from this PR as inspiration.

Thank you @guncebektas for your work! If you wanna continue to help out feel free to reach out!

File: packages/html-tools/templatetag.js [hunk 16]
```diff
@@ -1,29 +1,15 @@
...
+    return new TemplateTag();
+  }
 
-  if (props)
-    _assign(this, props);
+  if (props) Object.assign(this, props);
 }
 
-_assign(TemplateTag.prototype, {
+Object.assign(TemplateTag.prototype, {
```
[the remark this answers is not in the document]
Comment (StorytellerCZ) [reply packages/html-tools/templatetag.js:10]: Yes it does.

[the remark this answers is not in the document]
Comment (StorytellerCZ) [reply packages/html-tools/templatetag.js:10]: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_assign

**Comment by StorytellerCZ** [comment]
@guncebektas can you please take a look on @DanielDornhardt comments?
